src/extension/infer.rs

```diff
@@ -1133,6 +1160,502 @@ mod test {
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT], type_row![NAT]],
+                extension_delta: a.clone(),
+            },
+            FunctionType::new(vec![NAT], twoway(NAT)),
+        )?;
+
+        let exit = hugr.add_node_with_parent(
+            hugr.root(),
+            NodeType::open_extensions(ops::BasicBlock::Exit {
+                cfg_outputs: type_row![NAT],
+            }),
+        )?;
+
+        let mkpred = hugr.add_node_with_parent(
+            entry,
+            NodeType::open_extensions(make_opaque(
+                A,
+                FunctionType::new(vec![NAT], twoway(NAT)).with_extension_delta(&a),
+            )),
+        )?;
+
+        let bb0 = make_block(
+            &mut hugr,
+            root,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: bc.clone(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)),
+            make_opaque(
+                B,
+                FunctionType::new(vec![NAT], oneway(NAT)).with_extension_delta(&bc),
+            ),
+        )?;
+
+        let bb1 = make_block(
+            &mut hugr,
+            root,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT], type_row![NAT]],
+                extension_delta: b.clone(),
+            },
+            FunctionType::new(vec![NAT], twoway(NAT)),
+            make_opaque(
+                B,
+                FunctionType::new(vec![NAT], twoway(NAT)).with_extension_delta(&b),
+            ),
+        )?;
+
+        let bb10 = make_block(
+            &mut hugr,
+            root,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: c.clone(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)),
+            make_opaque(
+                C,
+                FunctionType::new(vec![NAT], oneway(NAT)).with_extension_delta(&c),
+            ),
+        )?;
+
+        let bb11 = make_block(
+            &mut hugr,
+            root,
+            ops::BasicBlock::DFB {
+                inputs: type_row![NAT],
+                other_outputs: type_row![],
+                predicate_variants: vec![type_row![NAT]],
+                extension_delta: c.clone(),
+            },
+            FunctionType::new(vec![NAT], oneway(NAT)),
+            make_opaque(
+                C,
+                FunctionType::new(vec![NAT], oneway(NAT)).with_extension_delta(&c),
+            ),
+        )?;
+
+        // Internal wiring for DFGs
```

Comment:
Move this up to where `entry`(_in,_out) and `mkpred` are declared

Reply:
(Like you are doing with `let entry =` in the `multi_entry` test below, tho I won't insist on hiding all these in a nested scope)